docs/porting/target/static_pinmap.md

```diff
@@ -0,0 +1,180 @@
+ - `UART`.
+ - `QSPI`.
+ - `CAN`.
+ 
+## Assumptions
+
+1. Provide types that will hold static pin maps for peripherals (`PWM`, `AnalogIn`, `AnalogOut`, `SPI`, `I2C`, `UART`, `QSPI`, `CAN`).
+2. Provide `xxx_init_direct(xxx_t *obj, static_pinmap_t *)` functions to HAL API (these functions won't use pin map tables).
+3. Provide additional constructors in the drivers layer, which will use the `xxx_init_direct(xxx_t *obj, static_pinmap_t*)` HAL functions.
+4. Provide default weak implementations of `xxx_init_direct(static_pinmap_t *)` functions. These functions call standard `xxx_init(xxx_t *obj, PinName, ...)` function (backward compatibility for targets that don't support the static pin map mechanism).
+5. Provide `constexpr` utility functions to lookup for pin mapping in compile time (requires C++14).
+6. Provide `constexpr` pin map tables in the header file.
+7. Provide macros for the pin map tables.
+8. Provide `STATIC_PINMAP_READY` macro in `PinNames.h`.
+
+## Implementing static pin map extension
+
+Most of the above points are already implemented. To make static pin map available on your platform, please perform the following steps: 
+
+- Provide implementation of `xxx_init_direct(xxx_t *obj, static_pinmap_t *)` function (which does not use pin map tables).
+   - `xxx_init()` uses pin map tables to determine the associated peripheral or function with the given pins, populates the pin map structure and calls void `xxx_init_direct()`.
+   - `xxx_init_direct()` performs peripheral initialization using given static pin map structure.
+
+   Example implementation:
+
+   ```
+   void xxx_init_direct(xxx_t *obj, const PinMap *pinmap)
+   {
+       obj->spi.instance = pinmap->peripheral;
+   
+       // pin out the xxx pins
+       pin_function(pinmap->pin, pinmap->function);
+       pin_mode(pinmap->pin, PullNone);
+   
+       // Some additional init code
+   }
+   
+   void xxx_init(xxx_t *obj, PinName pin)
+   {
+       int peripheral = (int)pinmap_peripheral(pin, PinMap_xxx);
+       int function = (int)pinmap_find_function(pin, PinMap_xxx);
+   
+       const PinMap static_pinmap = {pin, peripheral, function};
+   
+       xxx_init_direct(obj, &static_pinmap);
+   }
+   ```
+
+- Provide `constexpr` pin-map tables in the header file.
+
+   Move pin map tables from `PeripheralPins.c` to `PeripheralPinMaps.h` (create new file) and add `constexpr` specifier in the pin map table declarations.
+   
+   The tables are required in the header file, so constant expression utility functions can include and use them to find and return mapping without pulling the pin map table into the image.
+   
+   <span class="notes">**Note:** Please include the `<mstd_cstddef>` module, and use the `MSTD_CONSTEXPR_OBJ_11` macro instead of the `constexpr` specifier. This must be done for backward compatibility with the Arm 5 compiler, which does not support constant expressions. When you use the Arm 5 compiler, `MSTD_CONSTEXPR_OBJ_11` translates to `const`.</span>
```

Comment:
Do you only need to do this when using the Arm 5 compiler?

Reply:
This is required generally, but it is used to deal with `ARM 5` compiler. If we use `constexpr`, then we will get an error in the case of `ARM 5` compiler. If we use `MSTD_CONSTEXPR_OBJ_11` macro we will get in case of  `ARM 5` `const` (instead `constexpr` which will be used in all other cases).

Reply:
This wasn't actually anything to do with ARM C 5 - it was to handle the case where the header was going to be included from C code (inside the target code).

When we include it from the Mbed OS C++ code, we need to see it as `constexpr`, but if the target code includes it from C, it has to have it as `const`.

https://github.com/ARMmbed/mbed-os/pull/11399#discussion_r322608118

Reply:
Yes, but I remember that this was also related to ARMC5:

https://github.com/ARMmbed/mbed-os/pull/11399#discussion_r322610020

But this was about the `constexpr` functions and here we have `constexpr` tables.

I mixed this. Thanks for catching this.